### PR TITLE
Mac OSX: Added code signing capability with --osx-codesign-identity arg.

### PR DESCRIPTION
**Pull request by cculianu**
OSX only: if `--osx-codesign-identity` is specified on the command-line, then we will use the Xcode Developer Tools `codesign` utility to sign **all embedded binaries** (as well as the .app bundle itself). Note that this option may also be given to PKG and BUNDLE for finer-grained control as well, but it is recommended to just use the option on the command-line to ensure everything is signed with the same identity.

Let me know if this PR passes review properly and/or if anything needs to be modified.  Note that it outputs verbose output form the codesign utility which looks like this:

```
19541 INFO: Building PKG (CArchive) PKG-00.pkg
_struct.cpython-36m-darwin.so: replacing existing signature
_struct.cpython-36m-darwin.so: signed Mach-O thin (x86_64) [_struct.cpython-36m-darwin]
zlib.cpython-36m-darwin.so: replacing existing signature
zlib.cpython-36m-darwin.so: signed Mach-O thin (x86_64) [zlib.cpython-36m-darwin]
Python: replacing existing signature
Python: signed Mach-O thin (x86_64) [Python]

... etc ...
157934 INFO: Code signing BUNDLE with identity "Developer ID Application: My Test Identity"
MyTestApp.app: signed app bundle with Mach-O thin (x86_64) [org.mydomain.MyTestApp]
```
--

**Note:** It was always possible to code sign apps after building, but the embedded/collected binaries were unsigned.  

This facility adds the capability to code sign *deeply* so that the embedded binaries are signed as well (stuff like the Python interpreter, .so files that Python libs come with, embedded libs such as Qt, etc).

Apple has threatened/announced that in the future unsigned apps (and by extension unsigned embedded binaries) may not have as much access to OS facilities (the camera, microphone, etc) as signed apps have.

So, aside from that -- It's useful to be able to code sign apps as well as all the binaries embedded within the app. This is guaranteed to keep Mac's GateKeeper happy.

Also, this will be compatible with 'notarized apps' that is a feature Apple may require developers to comply with in the future. The Hardened Runtime introduced in macOS Mojave kills apps whose code signature gets invalidated (as far as I understand).

- Notarized apps must use the Hardened Runtime.
- In the future, all Developer ID signed apps must be notarized.

In short: There will be a point in the future where PyInstall either must implement this or be only usable for developers who do not sign their code at all.

### CODE REVIEW

**Comment by cculianu** [comment]
Note: Some of the Travis CI tests seem to have failed.  I don't believe this is related to this PR at all and could be some other issue with current develop branch, etc.

**Comment by cculianu** [comment]
Hi @htgoebel -- thanks so much for this detailed review.  I'm going to work through it later today and/or tomorrow and write back here if I have questions.

I just wanted to write back quickly right now re: funding.  No, I do not work for a company. I am working (as a volunteer) for the Electron Cash project (Bitcoin Cash wallet, Python based).  We lost our main funding source late last year after the hard fork with BSV and so most of the dev I am doing right now is volunteer based.  

Thanks again for your response and review and I will respond with commits and/or question(s) later...

File: PyInstaller/building/makespec.py [hunk 1]
```diff
@@ -280,6 +280,12 @@ def __add_options(parser):
                         'name for code signing purposes. The usual form is a hierarchical name '
                         'in reverse DNS notation. For example: com.mycompany.department.appname '
                         "(default: first script's basename)")
+    g.add_argument('--osx-codesign-identity', dest='codesign_identity',
+                   help='If this option is given, all binaries packaged in the app, including '
+                        'the app bundle itself, will be code signed using the ``codesign`` '
+                        'utility provided by Xcode Developer tools. Make sure the utility is in '
+                        'your PATH and that the identity specified is valid for code signing, eg '
+                        "'Developer ID Application: Your Name'")
```
[the remark this answers is not in the document]
Comment (cculianu) [reply PyInstaller/building/makespec.py]: It's 2 separate ideas.  

1. The `codesign` utility must be in your path.
2. The identity specified must be valid, e.g. "Developer ID Application: Your name"

Let me know if you really want me to rewrite this -- I figure terse is better here and to me the sentence makes sense. :)

File: PyInstaller/utils/osx.py [hunk 2]
```diff
@@ -103,3 +103,41 @@ def fix_exe_for_code_signing(filename):
     ## Write changes back.
     with open(exe_data.filename, 'rb+') as fp:
         exe_data.write(fp)
+
+def osx_code_sign(binary, identity, deep=None, verbose=True):
```
[the remark this answers is not in the document]
Comment (cculianu) [reply PyInstaller/utils/osx.py]: True. Will do.

File: PyInstaller/building/makespec.py [hunk 3]
```diff
@@ -280,6 +280,12 @@ def __add_options(parser):
                         'name for code signing purposes. The usual form is a hierarchical name '
                         'in reverse DNS notation. For example: com.mycompany.department.appname '
                         "(default: first script's basename)")
+    g.add_argument('--osx-codesign-identity', dest='codesign_identity',
+                   help='If this option is given, all binaries packaged in the app, including '
+                        'the app bundle itself, will be code signed using the ``codesign`` '
+                        'utility provided by Xcode Developer tools. Make sure the utility is in '
+                        'your PATH and that the identity specified is valid for code signing, eg '
```
[the remark this answers is not in the document]
Comment (cculianu) [reply PyInstaller/building/makespec.py]: 👍

File: PyInstaller/building/utils.py [hunk 4]
```diff
@@ -32,6 +32,7 @@
 from ..utils import misc
 from ..utils.misc import load_py_data_struct, save_py_data_struct
 from .. import log as logging
+from ..utils.osx import osx_code_sign
```
[the remark this answers is not in the document]
Comment (cculianu) [reply PyInstaller/building/utils.py]: Ok, done in my version (will force push later when done).

File: PyInstaller/utils/osx.py [hunk 5]
```diff
@@ -103,3 +103,41 @@ def fix_exe_for_code_signing(filename):
...
+    """ Code sign binary (.so, .dylib, or .app) using identity.
+
+    Relies on the codesign utility which must be in PATH.
+
+    binary
+        (str) Full path to the binary or .app to be signed.
+    identity
+        (str) The code signing identity to use (codesign -s option).
```
[the remark this answers is not in the document]
Comment (cculianu) [reply PyInstaller/utils/osx.py]: Hmm. No.. it means the exact "string" that you would give to `codesign -s`. I'll reword this to be clearer.

File: PyInstaller/utils/osx.py [hunk 6]
```diff
@@ -103,3 +103,41 @@ def fix_exe_for_code_signing(filename):
...
+        deep = True
+    deep = '--deep' if deep else ''
+    verbose = '-v' if verbose else ''
+
+    d = os.path.dirname(binary)
+    saved_dir=None
+    if d:
+        # switch to directory of the binary so codesign verbose messages don't
+        # include long path
+        saved_dir = os.path.abspath(os.path.curdir)
```
[the remark this answers is not in the document]
Comment (cculianu) [reply PyInstaller/utils/osx.py]: Ok. 👍

File: PyInstaller/utils/osx.py [hunk 7]
```diff
@@ -103,3 +103,41 @@ def fix_exe_for_code_signing(filename):
...
+        os.chdir(d)
+        binary = os.path.basename(binary)
+    if 0 != os.system("codesign {} -f {} -s '{}' '{}'".format(verbose, deep,
+                                                              identity, binary)):
+        raise SystemExit('Error: Failed to code sign binary {} using identity '
+                         '"{}". Please ensure that the ``codesign`` utility '
+                         'is in your PATH and that the identity specified is '
+                         'valid for signing.'.format(binary, identity))
+    if saved_dir:
+        os.chdir(saved_dir)
```
[the remark this answers is not in the document]
Comment (cculianu) [reply PyInstaller/utils/osx.py]: Ok, fair enough. Better safe than sorry to always ensure it gets done that way, I agree. Done in my local repo.

File: PyInstaller/building/api.py [hunk 8]
```diff
@@ -181,6 +184,11 @@ def __init__(self, toc, name=None, cdict=None, exclude_binaries=0,
         self.exclude_binaries = exclude_binaries
         self.strip_binaries = strip_binaries
         self.upx_binaries = upx_binaries
+        if is_darwin:
+            from ..config import CONF
+            self.codesign_identity = codesign_identity if codesign_identity is not None else CONF['codesign_identity']
```
[the remark this answers is not in the document]
Comment (cculianu) [reply PyInstaller/building/api.py]: I'm not exactly sure how else to pass command-line arguments forward as "defaults" to the various classes (EXE, PKG, osx.BUNDLE, etc).  Can you illuminate me on how this would be done?

I don't want users in the .spec files to have to manually enter:
```
EXE(...,codesign_identity="Bla bla")
#...
BUNDLE(...,codesign_identity="Bla bla bla")
# ..
```

When they could just invoke pyinstaller with `pyinstaller --osx-codesign-identity XXX`.

I'll go back and read the sources more and try and figure it out.  But any pointers on how to have "global defaults" that come from the command-line here would be really helpful.

Let me know!

File: PyInstaller/building/osx.py [hunk 9]
```diff
@@ -14,6 +14,7 @@
 from .api import EXE, COLLECT
 from .datastruct import Target, TOC, logger
 from .utils import _check_path_overlap, _rmtree, add_suffix_to_extensions, checkCache
+from ..utils.osx import osx_code_sign
```
[the remark this answers is not in the document]
Comment (cculianu) [reply PyInstaller/building/osx.py]: 👍

File: PyInstaller/utils/osx.py [hunk 10]
```diff
@@ -103,3 +103,41 @@ def fix_exe_for_code_signing(filename):
...
+    Relies on the codesign utility which must be in PATH.
+
+    binary
+        (str) Full path to the binary or .app to be signed.
+    identity
+        (str) The code signing identity to use (codesign -s option).
+    deep
+        (bool) Specify whether to use the codesign --deep option. If not
+        specified, will auto-detect based on whether binary is a .app directory
+        or a stand-alone file.
```
[the remark this answers is not in the document]
Comment (cculianu) [reply PyInstaller/utils/osx.py]: Ok.. the binary could theoretically also be a `.framework` which is also a directory (in some future use of this function -- not as pyinstaller currently operates).  I clarified the working in my local repo.

File: PyInstaller/utils/osx.py [hunk 11]
```diff
@@ -103,3 +103,41 @@ def fix_exe_for_code_signing(filename):
...
+    verbose
+        (bool) Specify whether codesign output should be verbose.
+    """
+    if deep is None and os.path.isdir(binary):
+        deep = True
+    deep = '--deep' if deep else ''
+    verbose = '-v' if verbose else ''
+
+    d = os.path.dirname(binary)
```
[the remark this answers is not in the document]
Comment (cculianu) [reply PyInstaller/utils/osx.py]: Ok, done.

File: PyInstaller/utils/osx.py [hunk 12]
```diff
@@ -103,3 +103,41 @@ def fix_exe_for_code_signing(filename):
...
+    saved_dir=None
+    if d:
+        # switch to directory of the binary so codesign verbose messages don't
+        # include long path
+        saved_dir = os.path.abspath(os.path.curdir)
+        os.chdir(d)
+        binary = os.path.basename(binary)
+    if 0 != os.system("codesign {} -f {} -s '{}' '{}'".format(verbose, deep,
+                                                              identity, binary)):
```
[the remark this answers is not in the document]
Comment (cculianu) [reply PyInstaller/utils/osx.py]: Ah ok.  Not a problem.

**Comment by cculianu** [comment]
Hi @htgoebel ,

Thanks so much for your very clear and detailed instructions.  I'm working on an amended commit in my local repository and when it's ready I'll force push it.  You have been so helpful with all the clear directions you have sent me that I can't thank you enough.  I did about 80%-95% of what you recommended already with 1 minor question remaining regarding this: https://github.com/pyinstaller/pyinstaller/pull/3991#discussion_r249793815.

- Basically, I'm not exactly sure how one may "pass forward" the command-line options to the various submodules (EXE, PKG, BUNDLE, etc) -- which is why I resorted to the `CONF` approach.  Do let me know how this is normally done in the sources if you can.  In the meantime I'll continue to research the sources.

**Comment by cculianu** [comment]
@htgoebel -- I force pushed another commit that does (almost) everything you recommended:

- Commit message is to spec
- Added documentation in the manual
- Lint fixups
- All you code recommendations (with 1 exception, see below)
- Changelog entry added
- Added code to _GUTS for classes where codesign_identity is used.

Only extant issue is this one, which I need more guidance/discussion on: https://github.com/pyinstaller/pyinstaller/pull/3991#discussion_r249793815

**Comment by cculianu** [comment]
Note that once we resolve the CONF issue -- I will go back and re-do the logic/idioms for how the `self.codesign_identity` attribute is computed, as per your recommendations.. so either way this is not the final force-pushed commit. 

(I'm also curious to see if any lint errors still exist, hence why I pushed this now).

**Comment by cculianu** [comment]
@htgoebel So... ping?

**Comment by cbenhagen** [comment]
@htgoebel is there anything to do to get this merged?

**Comment by cculianu** [comment]
@htgoebel Apologies if you are busy -- I'm willing to do whatever it takes to fix this up to your liking.  I had a few questions for your from your previous comments.  Do let me know if/what I can do. Thanks.

File: PyInstaller/utils/osx.py [hunk 13]
```diff
@@ -103,3 +103,59 @@ def fix_exe_for_code_signing(filename):
...
+            os.chdir(binary_dir)
+            saved_cwd = _saved  # chdir success, save for `finally:` below
+            binary = os.path.basename(binary)
+        args = ["codesign", "-f", "-s", identity, binary]
+        fargpos = 1
+        if verbose:
+            args.insert(fargpos, verbose)
+            fargpos += 1
+        if deep:
+            args.insert(fargpos + 1, deep)
```
Comment (htgoebel) [inline PyInstaller/utils/osx.py]: Why not just
```
args = ["codesign", "-f", "-s", identity, binary]
if verbose:
    args.append("-v")
if deep:
    args.append("--deep")
```
(I assume options can follow args on OSX, too) or at least
```
args = ["codesign", "-f", "-s", identity]
verbose:
    args.append("-v")
if deep:
    args.append("--deep")
args.append(binary)
```

File: PyInstaller/utils/osx.py [hunk 14]
```diff
@@ -103,3 +103,59 @@ def fix_exe_for_code_signing(filename):
...
+        (bool) Specify whether to use the codesign --deep option. If not
+               specified, will auto-detect based on whether ``binary`` is a
+               directory (such as a .app or .framework) or a stand-alone file.
+    verbose
+        (bool) Specify whether codesign output should be verbose.
+    """
+    if deep is None and os.path.isdir(binary):
+        deep = True
+    deep = '--deep' if deep else ''
+    verbose = '-v' if verbose else ''
```
Comment (htgoebel) [inline PyInstaller/utils/osx.py]: No need for these lines anymore since the `if` is done below.

File: PyInstaller/utils/osx.py [hunk 15]
```diff
@@ -103,3 +103,59 @@ def fix_exe_for_code_signing(filename):
...
+
+    binary_dir = os.path.dirname(binary)
+    saved_cwd = None
+    try:
```
Comment (htgoebel) [inline PyInstaller/utils/osx.py]: You can save this outer level or `try` by moving the `chdir` into the inner `try` and the `finally` below the inner `try`. Most of this code should not fail (except in case of an memory error, which will terminate PyInstaller anyway).

File: PyInstaller/utils/osx.py [hunk 16]
```diff
@@ -103,3 +103,59 @@ def fix_exe_for_code_signing(filename):
...
+    verbose
+        (bool) Specify whether codesign output should be verbose.
+    """
+    if deep is None and os.path.isdir(binary):
+        deep = True
+    deep = '--deep' if deep else ''
+    verbose = '-v' if verbose else ''
+
+    binary_dir = os.path.dirname(binary)
+    saved_cwd = None
```
Comment (htgoebel) [inline PyInstaller/utils/osx.py]: You can assign to `saved_cwd` directly here. Even if `chdir(binary_dir)` would fail, running `os.chdir(saved_cwd)` in `finally:` does no harm. Or did I miss some case?

(And in the `finally`-claus you can save the `if saved_cwd:` test.)

File: PyInstaller/utils/osx.py [hunk 17]
```diff
@@ -103,3 +103,59 @@ def fix_exe_for_code_signing(filename):
...
+            args.insert(fargpos, verbose)
+            fargpos += 1
+        if deep:
+            args.insert(fargpos + 1, deep)
+
+        err_msg, result = '', None
+        try:
+            result = exec_command_rc(*args)
+        except OSError as e:
+            err_msg = " (Error was: " + repr(e) + ")"
```
Comment (htgoebel) [inline PyInstaller/utils/osx.py]: `" (Error was: %r)" % e`

File: PyInstaller/utils/osx.py [hunk 18]
```diff
@@ -103,3 +103,59 @@ def fix_exe_for_code_signing(filename):
...
+        if 0 != result:
```
Comment (htgoebel) [inline PyInstaller/utils/osx.py]: Please write this `result != 0`.

File: doc/usage.rst [hunk 19]
```diff
@@ -418,6 +418,24 @@ and for more detail, the `Apple code signing overview`_ technical note).
 You can add other items to the :file:`Info.plist` by editing the spec file;
 see :ref:`Spec File Options for a Mac OS X Bundle` below.
 
+Code Signing Embedded Binaries
+-------------------
```
Comment (htgoebel) [inline doc/usage.rst]: Invalid reStrcutre Text: "underline" must be at least as long as the text.

File: doc/usage.rst [hunk 20]
```diff
@@ -418,6 +418,24 @@ and for more detail, the `Apple code signing overview`_ technical note).
...
+
+You may also optionally specify that |PyInstaller| code sign all binaries which
+get embedded into your built application (such as the Python interpreter, its
+.so files, .dylib files for packages such as PyQt, etc). To enable code signing
+of embedded binaries, use the ``--osx-codesign-identity`` command-line option
+(or ``codesign_identity=`` kw arg). This facility requires the ``codesign``
```
Comment (htgoebel) [inline doc/usage.rst]: (or `codesign_identity=` ~~kw~~ arg**ument** to `BUNDLE`in the spec-file)

File: doc/usage.rst [hunk 21]
```diff
@@ -418,6 +418,24 @@ and for more detail, the `Apple code signing overview`_ technical note).
...
+utility (which comes installed with a full install of Xcode) be present in your
```
Comment (htgoebel) [inline doc/usage.rst]: **to** be

File: doc/usage.rst [hunk 22]
```diff
@@ -418,6 +418,24 @@ and for more detail, the `Apple code signing overview`_ technical note).
...
+PATH.
+
+The identity argument must be a valid code signing identity installed on the
```
Comment (htgoebel) [inline doc/usage.rst]: I'd move this paragraph below the next one.

File: doc/usage.rst [hunk 23]
```diff
@@ -418,6 +418,24 @@ and for more detail, the `Apple code signing overview`_ technical note).
 You can add other items to the :file:`Info.plist` by editing the spec file;
 see :ref:`Spec File Options for a Mac OS X Bundle` below.
 
+Code Signing Embedded Binaries
+-------------------
+
+You may also optionally specify that |PyInstaller| code sign all binaries which
```
Comment (htgoebel) [inline doc/usage.rst]: … |PyInstaller| ~~code~~ sign**s** …

File: news/3991.feature.rst [hunk 24]
```diff
@@ -0,0 +1,2 @@
+Addition of embedded binary code signing capability for MacOS using the
+``--osx-codesign-identity`` command-line option.
```
Comment (htgoebel) [inline news/3991.feature.rst]: I suggest:
```
(MacOS) Added ``--osx-codesign-identity`` command-line option for code signing embedded binaries,
```
OTOH, this also signs the app, doesn't it?

File: PyInstaller/building/api.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (htgoebel) [reply PyInstaller/building/api.py]: So we have different cases:

* for `PKG` (which is not exposed in the spec-file): the function argument should be used (passed by `EXE`, as already done [here](https://github.com/pyinstaller/pyinstaller/blob/15ad5b7d805d0905d0ce8cc5b01d19e992796d85/PyInstaller/building/api.py#L447)

* for `EXE` and `BUNDLE` we first need to decide 1) whether we should require the identity to be specified on the command line for actually signing or 2) whether i should be possible to specify the ID in the spec-file. **What do you think**?

  * **If 1**:   Both `EXE` and `BUNDLE` should **not** accept a command line argument, but take the value from `CONF` only`. This would avoid one is specifying two conflicting values.

  * **if 2**: The ID could be put into the .spec-file by adding the appropriate line to the templates `PyInstaller/building/templates.py` and `PyInstaller/building/makespec.py`.

**Comment by cbenhagen** [comment]
@cculianu Any news?

**Comment by cculianu** [comment]
Sorry I haven't really had time to work on this.  I may find time soon but I lost a bit of momentum with it.. and have just been using my own patched private version for my own project.  Let me know though if there is serious interest in merging it and I can resume..

**Comment by phw** [comment]
Just for me to understand: How is this different then running 

```
codesign --keychain $KEYCHAIN_PATH --verify --verbose --deep --sign "$CERTIFICATE_NAME" 'My App.app'
```
Shouldn't the above also codesign all included binaries due to the deep option?

>     * Notarized apps must use the Hardened Runtime.

As I understand it this would require the `--options runtime` parameter to the codesign call (available since macOS 10.14), which does not happen here. How would you deal with this?

**Comment by cculianu** [comment]
> Just for me to understand: How is this different then running
> 
> ```
> codesign --keychain $KEYCHAIN_PATH --verify --verbose --deep --sign "$CERTIFICATE_NAME" 'My App.app'
> ```
> 
> Shouldn't the above also codesign all included binaries due to the deep option?
> 
> > ```
> > * Notarized apps must use the Hardened Runtime.
> > ```
> 
> As I understand it this would require the `--options runtime` parameter to the codesign call (available since macOS 10.14), which does not happen here. How would you deal with this?

Well PyInstaller creates an app within an app.  So if you codesign just the `.app`  it produces, you aren't codesigning the binaries that the app EMBEDS.

When PyInstaller executes it writes to a tmp dir in $TMP/_MEIxxx.  It extracts a bunch of the binaries packaged into your app there and then executes your app AGAIN from there.

Those embedded binaries need to be signed.  And there is no way to sign them by just signing the `.app` PyInstaller produces unless you use this patch.

**Comment by aliblackwell** [comment]
Is there anything I can do to help this towards full merge? @caiguanhao if I were to resolve the conflicts in another branch for you to merge in, would that be helpful?

**Comment by xylix** [comment]
@cculianu Sorry to ping you from such an old thread, but I just wanted to make sure:

[Pyinstaller's docs](https://pyinstaller.readthedocs.io/en/stable/usage.html#rarely-used-special-options) make it sound like using a tmpdir solution is optional, and mandatory only when compiling the program to a single-file executable. Did you every try / was it an option to use a non-single file executable (so just let pyinstaller pack all the libs and the main executable into the .app directory, without combining them)?

**Comment by cculianu** [comment]
@xylix I haven't even tried this because I did not know that was possible.  I think that approach would solve a lot of problems for both this and app startup times (Catalina + single-file executable makes the startup time on Catalina take MINUTES -- because it has to phone home for virus check or some nonsense!!).

I will try this. Thanks for the tip.

**Comment by xylix** [comment]
We've also been looking at how pyinstaller packages stuff and trying to make our macOS builds work smoothly in https://github.com/ActivityWatch/activitywatch .

Glad to be of use to someone else also working with the pains of getting macOS app packaging work properly :). Good luck.

**Comment by phw** [comment]
For MusicBrainz Picard we have the macOS builds working with code signing and notarization. And yes, we don't use the single-file approach, that's also why I was confused why special handling for this is necessary. If you create a normal .app package signing is possible with the codesign command.

If you want to look how we do this, here are the relevant files:

Running PyInstaller and use codeisgn: https://github.com/metabrainz/picard/blob/master/scripts/package/macos-package-app.sh

Notarize the app package: https://github.com/metabrainz/picard/blob/master/scripts/package/macos-notarize-app.sh

Everything is automatically done in our CI using Github Actions: https://github.com/metabrainz/picard/blob/master/.github/workflows/package.yml#L24

**Comment by cculianu** [comment]
Thank you for this! I had no idea that was an option.  I appreciate it.

**Comment by rokm** [comment]
I'm closing this as we ended up implementing practically the same functionality as part of adding support for M1 (#5581), where we need to resign the collected binaries anyway.
